feat(auth): Sync classic/API PAS plugin cookies

rpatterson opened the pull request:
Build on #1141 to flesh out the rest of the PAS plugin interfaces that the JWT plugin
should support, namely those methods that handle log in and log out.  With these changes
logging in to or out of Plone classic or the Volto UI does the same in the other.  This
should result in a much less surprising user experience for those users that will end up
using both interfaces and makes the JWT plugin more correct and complete.  See also [the
related Volto PR](https://github.com/plone/volto/pull/2687).

Regarding the use case, I understand the intention is to move away from Plone classic as
a supported UI.  I think it would be a painful mistake to imagine that's going to be a
reality for all important use cases and user classes anytime soon.  We should expect
Volto to be the UI for an increasingly major portion of use cases over time but allow it
time to get to all of them.  Personally, I wouldn't plan on classic being something we
could stop supporting for a couple of major versions of Plone.

mister-roboto commented:
@rpatterson thanks for creating this Pull Request and help improve Plone!

To ensure that these changes do not break other parts of Plone, the Plone test suite matrix needs to pass.

Whenever you feel that the pull request is ready to be tested, either start all jenkins jobs pull requests by yourself, or simply add a comment in this pull request stating:

```
@jenkins-plone-org please run jobs
```

With this simple comment all the jobs will be started automatically.

Happy hacking!

rpatterson commented:
Apologies, I went on auto pilot and stopped running the full  test suite locally.  I can see these errors locally and I'm working on them now.

rpatterson commented:
@jenkins-plone-org please run jobs

tiberiuichim commented on src/plone/restapi/pas/plugin.py:
```diff
@@ -39,7 +41,13 @@ def addJWTAuthenticationPlugin(self, id_, title=None, REQUEST=None):
+    IAuthenticationPlugin,
+    IChallengePlugin,
+    IExtractionPlugin,
+    plugins_ifaces.ICredentialsUpdatePlugin,
```
I think the already existing style of imports should be kept.

rpatterson replied:
I find the `module.attribute` form of reference to imported code to be much more readable because I can see where the imported code comes from where it's code is used.  But not a hill I'll die on today.  Done.

tiberiuichim commented on src/plone/restapi/pas/plugin.py:
```diff
@@ -127,6 +146,51 @@ def authenticateCredentials(self, credentials):
         return (userid, userid)
 
+    @security.private
+    def updateCredentials(self, request, response, login, new_password):
```
So, if I understand correctly, you're making this plugin act when logging in from Plone Classic, in addition to the other PAS plugins, right? Seems ok to me.

rpatterson replied:
Pretty much.  Only a little better because it obeys the PAS plugin interfaces configuration.  So site administrators can decide which plugins should "participate" on login/logout "events".

rpatterson commented:
@jenkins-plone-org please run jobs

rpatterson commented:
@jenkins-plone-org please run jobs

rpatterson commented:
I believe I've addressed all your feedback, @tisto.  LMK if I missed anything or if there's anything else I can do to move this forward.  Thanks for the review!

rpatterson commented:
I just realized that some of my previous local development build improvements make it
pretty easy to create a Zope and Plone instance from before my various auth PRs and then
test it under all my PRs.  Having done so, it confirms that at least some sort of
[upgrade step](https://github.com/plone/plone.restapi/pull/1304#issue-1090167421) is
required.  Specifically, because token generation has been moved into `updateCredentials(...)`
 we need an upgrade step that enables the JWT token plugin for that PAS plugin
 interface.  IOW, this is *not* ready for merge.

wesleybl commented:
@rpatterson Just out of curiosity, what's your use case? Only authenticated access to Plone Classic? Is there any situation where the anonymous accesses Plone Classic? Does plone api return HTML to Volto?

rpatterson commented:
> @rpatterson Just out of curiosity, what's your use case?

I'm on the EEA team, so I'm working under the direction of their leadership.  But as I understand it, the use case is to have the experience of using Volto, Plone classic, and the ZMI to be seamless in terms of logging in and out.  IOW, it doesn't matter where you log in, you'll be logged in everywhere and ditto for logging out.

> Only authenticated access to Plone Classic? Is there any situation where the anonymous accesses Plone Classic? Does plone api return HTML to Volto?

I don't really follow these questions.  Can you clarify?

wesleybl commented:
What I was wondering, is what is your site's workflow? Do content managers authenticate to Plone Classic to register content (Page, News...)? After this content consumed by Volto to anonymous users?

tiberiuichim commented:
@wesleybl @rpatterson Although I'm not on the EEA team which is currently handling this feature request, I think I can add some insight. We currently have a Volto website (eea.europa.eu/ims) running as a subfolder of a Plone 4 classic website. The intention is to have editors seamless authentication between the classic Plone and the Volto part. I think the ZMI auth is a nice to have, but I may be wrong.

rpatterson commented:
@jenkins-plone-org please run jobs

rpatterson commented:
> ...it confirms that at least some sort of upgrade step is required. ... IOW, this is not ready for merge.

I just pushed the upgrade step and related changes.  So this should be ready for review and merge.

rpatterson commented:
> I just pushed [the upgrade step and related changes](https://github.com/plone/plone.restapi/commit/df65b2e7fd551929213dd303396e340eaa0ded6a). So this should be ready for review and merge.

Don't suppose anyone can review this?  Is there anything blocking review here?  Sorry to pester, but... I'm pestering.  ;-)  CC: @sneridagh @tiberiuichim @wesleybl

wesleybl commented on src/plone/restapi/setuphandlers.py:
```diff
@@ -31,19 +26,26 @@ def getNonInstallableProducts(self):  # pragma: no cover
+                [
+                    "IAuthenticationPlugin",
+                    "IExtractionPlugin",
+                    "ICredentialsUpdatePlugin",
+                    "ICredentialsResetPlugin",
```
Should these plugins really be enabled by default? Isn't this a specific use case? For those who don't care about integrated authentication, wouldn't that be an unnecessary burden (less performance)?

rpatterson replied:
> Should these plugins really be enabled by default?

Yes.  The purpose of these changes and this PR is to keep login/logout state in sync between plugins.  Those interfaces are the ones that are responsible for that.

> Isn't this a specific use case?

This is the common use case.  A user who is accessing Plone from multiple interfaces, such as Plone Classic and Volto, and those interfaces use different PAS plugins, such as a cookie plugin and a JWT token plugin, that user has every reason to expect that logging out of one also logs them out of the other.  This is the least surprising behavior and as such should be the default.

> For those who don't care about integrated authentication

We can be grateful to PAS for making that a simple matter of deactivating the plugins for those interfaces in that specific use case.  ;-)

> wouldn't that be an unnecessary burden (less performance)?

Login/logout are rare events and as such not a significant impact on the user experience of performance.  Both of these operations for both of these plugins are also very cheap and as such wouldn't impact performance anyways.  Nor do I know of *any* plugin for which these operations are other than very cheap.

rpatterson replied:
Feel free to unresolve with details if this still is blocking merge!

wesleybl replied:
Well I'll mark it as unresolved so other people can see your points. But I don't think this is a blocker. I just wish other people would give their opinions too.

avoinea commented:
@rpatterson 

I added this branch to my existing Plone 6 setup https://github.com/eea/eea-website-backend/tree/master/develop

And I get this error when accessing http://localhost:8080

```
2022-02-10 18:36:39 ERROR [Zope.SiteErrorLog:35][waitress-2] ComponentLookupError: http://localhost:8080/manage_main
Traceback (innermost last):
  Module ZPublisher.WSGIPublisher, line 167, in transaction_pubevents
  Module ZPublisher.WSGIPublisher, line 376, in publish_module
  Module ZPublisher.WSGIPublisher, line 255, in publish
  Module ZPublisher.BaseRequest, line 619, in traverse
  Module Products.PluggableAuthService.PluggableAuthService, line 244, in validate
  Module Products.PluggableAuthService.PluggableAuthService, line 603, in _extractUserIds
  Module plone.restapi.pas.plugin, line 133, in authenticateCredentials
  Module plone.restapi.pas.plugin, line 214, in _decode_token
  Module zope.component._api, line 165, in getUtility
zope.interface.interfaces.ComponentLookupError: (<InterfaceClass plone.keyring.interfaces.IKeyManager>, '')
```

rpatterson commented:
> I added this branch to my existing Plone 6 setup https://github.com/eea/eea-website-backend/tree/master/develop `source.ini`
> 
> And I get this error when accessing http://localhost:8080
> 
> ```
...
> zope.interface.interfaces.ComponentLookupError: (<InterfaceClass plone.keyring.interfaces.IKeyManager>, '')
> ```

Did you run [the upgrade step](https://github.com/plone/plone.restapi/pull/1303#issuecomment-1029474261), @avoinea?

avoinea commented:
> Did you run [the upgrade step](https://github.com/plone/plone.restapi/pull/1303#issuecomment-1029474261), @avoinea?

As I can't access anything, including `/manage_main` I'm not sure how a common site admin will be able to run the upgrade step :thinking:

rpatterson commented:
@jenkins-plone-org please run jobs

rpatterson commented:
> As I can't access anything, including `/manage_main` I'm not sure how a common site admin will be able to run the upgrade step thinking

Hmm, I don't understand.  The exception should only happen once the PAS interfaces are activated but that's exactly what the upgrade step does.  Can you try [the steps I did to test the upgrade step](https://github.com/plone/plone.restapi/commit/03dd65a84c87cf84afad9ede4d70694242030742) and see if you can reproduce what you observe in the Volto buildout.  Can elaborate on your test steps and how they differ, @avoinea?

rpatterson commented:
@jenkins-plone-org please run jobs

rpatterson commented:
Looks like [Jenkins is broken](https://github.com/plone/Products.PlonePAS/pull/66#issuecomment-1039531403), I'm seeing the same thing here.  Anyone have any insight?

avoinea commented:
@rpatterson These are the steps to reproduce:

1. Install https://github.com/eea/eea-website-backend/tree/master/develop#install
3. Create a Plone site
4. Stop Zope instance
4. Add `plone.restapi` to `sources.ini` (branch `fix-unify-auth-logout`)
5. Run `make develop`
6. Run `make start`
7. Go to http://localhost:8080
7. You'll get the `ComponentLookup` error

It seems you can still go to http://localhost:8080/Plone and run the upgrade step. But the error is still there after the upgrade on Zope root.

rpatterson commented:
> @rpatterson These are the steps to reproduce:

And the steps I used?  How is the interface being activated before the upgrade step in your test bed?

rpatterson commented:
@jenkins-plone-org please run jobs

rpatterson commented:
> Do I have to also checkout other packages from different branches?

Oh yeah, sorry, @avoinea!  When I first started stitching togetherr PRs across the repos involved in Volto, I was maintaining and submitting PRs in `plone/volto` as well that included the checkouts required to test it all together in that repo.  I was told we don't handle it that way, I stopped maintaining such `plone/volto` branches.  I've refreshed it all and pushed to [the `fix-unify-auth` branch](https://github.com/plone/volto/commits/fix-unify-auth), so you should be able to use that to test or as a source for how to test properly in that other project.

rpatterson commented:
@jenkins-plone-org please run jobs

rpatterson commented:
I've now run [the failing Jenkins job](https://jenkins.plone.org/job/pull-request-6.0-3.9/905/console) 3 times and it's failing at the same place every time:

    ==============================================================================
    Test Overlays :: These tests are just testing the overlay behavior not the ...
    ==============================================================================
    Scenario: Set default content item of a folder overlay opens          | FAIL |
    Setup failed:
    WebDriverException: Message: chrome not reachable
    
    
    Also teardown failed:
    No browser is open.
    ------------------------------------------------------------------------------
    Test Overlays :: These tests are just testing the overlay behavior... | FAIL |
    1 critical test, 0 passed, 1 failed
    1 test total, 0 passed, 1 failed
    ==============================================================================
    Output:  /home/jenkins/workspace/pull-request-6.0-3.9/parts/test/test_overlays/Scenario_Set_default_content_item_of_a_folder_overlay_opens/output.xml
    
    
    
    Failure in test Scenario Set default content item of a folder overlay opens (test_overlays.robot)
    Traceback (most recent call last):
      File "/srv/python3.9/lib/python3.9/unittest/case.py", line 59, in testPartExecutor
        yield
      File "/srv/python3.9/lib/python3.9/unittest/case.py", line 592, in run
        self._callTestMethod(testMethod)
      File "/srv/python3.9/lib/python3.9/unittest/case.py", line 550, in _callTestMethod
        method()
      File "/home/jenkins/.buildout/eggs/cp39/robotsuite-2.2.1-py3.9.egg/robotsuite/__init__.py", line 475, in runTest
        assert last_status == 'PASS', last_message
    AssertionError: Setup failed:
    WebDriverException: Message: chrome not reachable
    
    
    Also teardown failed:
    No browser is open.

I've tested that overlay in Plone classic in my local development environment and it seems to be working just fine.  I can't imagine how this failure could be related to the changes here and I've been told before that the robot tests are intermittent.  So can I either get some help understanding how this is related to the changes here or can we ignore that failure and proceed?  CC: @avoinea

rpatterson commented:
> I've now run [the failing Jenkins job](https://jenkins.plone.org/job/pull-request-6.0-3.9/905/console) 3 times and it's failing at the same place every time:

Apparently, [the 4th time](https://jenkins.plone.org/job/pull-request-6.0-3.9/909/) is the charm.  This kind of intermittent CI false-negative is a big problem for an open source community.  Personally, I think intermittent false failures/errors/negatives in CI should be immediately "commented out" or otherwise skipped and some part of whatever resources the community or governance has to work with should be dedicated to resolving intermittent CI issues continuously over the longer term.

At any rate, AFAIK, this is ready to finish review and merge.  LMK if I'm missing anything?  CC: @avoinea @rpatterson tiberiuichim @wesleybl @sneridagh

rpatterson commented:
Is anyone able to finish reviewing this or help me draw attention to this to help finish review?  CC: @avoinea @tiberiuichim @wesleybl @sneridagh

wesleybl commented:
@rpatterson When I already have the Volto site open in the browser and I authenticate in the Plone Site, when I return to the Volto Site, I do not see the authenticated site. Only after I press F5 on Volto site do I see the authenticated site. Could anything be done?

jensens commented:
@wesleybl and vice versa, right? Unless there's no magic notification system like with websockets I fear this is not possible. And I would not expect this.

tiberiuichim commented:
Sessionstorage (or localstorage) would help here, but not without JS support to poll that storage periodically.

wesleybl commented:
> @wesleybl and vice versa, right?

@jensens vice versa this does not happen. If I have a Volto tab and a Plone tab, without being authenticated to either, after I authenticate to Volto and navigate in Plone, I see Plone authenticated.  The Plone site checks the cookie on every request.

> And I would not expect this.

I would expect for this. I know several systems that have synchronized authentication, and it is not necessary to press F5 to see the authenticated system. Many users don't know what the F5 does. These users may find that synchronized authentication does not work in some situations.

Anyway, I don't think this would be a block. Especially if it harms the Volto's performance. But it would be a plus.

rpatterson commented:
> @wesleybl and vice versa, right? Unless there's no magic notification system like with websockets I fear this is not possible. And I would not expect this.

Agreed, @jensens.  This *might* be a nice-to-have feature for *some* use cases but it's also a significant increase in complexity so I'm not sure how I'd end up voting if reviewing such a PR.  Regardless, such behavior should be evaluated in a separate PR and I think this PR is an improvement and results in more correct and less-surprising behavior as is without such additional gloss.

At any rate, everything is green in this PR.  Should I merge it?  Is that something someone else should do and if so, who?  Thanks!

CC: @avoinea

jensens commented:
I would like to have this merged by @tisto or @sneridagh as it is usually done.

tisto commented:
@rpatterson if I am not mistaken we agreed, at a @plone/volto-team meeting, that EEA will test this extensively on their live site before we merge this. Did that happen?

rpatterson commented:
> @rpatterson if I am not mistaken we agreed, at a @plone/volto-team meeting, that EEA will test this extensively on their live site before we merge this. Did that happen?

That's not my understanding or recollection.  How about other EEA folks?  I know @alecghica was at that meeting?  I think @avoinea was too?  I can't recall how well suited the EEA live site deployment story is to using git checkouts with specific branches, nor to run upgrade steps (and install additional profiles for the Zope root work in the other PR).  I also know the EEA teams are quite busy, so I would be surprised if this was something that could be done in a timely way.

I'm concerned about bit rot.  Also, my time dedicated to this unified auth work is running out sometime this month.  After that, if I'm able to put any time into this at all, I'll probably have to choose between either shepherding PRs through or making any actual code changes.  Given all that, I think there's significant risk that all this work will be wasted if such testing is a requirement for merging.  So for my part, I'd rather merge sooner and risk having to get a bug fix out the door than wait for perfect safety to get this merged.  What do others think?

avoinea reviewed (approved):
LGTM. 

@rpatterson @mauritsvanrees Just make sure the upgrade step hits plone.app.upgrade / the [ADDON_LIST of core profiles with upgrades](https://github.com/plone/Products.CMFPlone/issues/3453)

rpatterson commented:
> LGTM.

Does that mean you've done some sort of additional testing to address @tisto's question, @avoinea?  [Ditto for #1304](https://github.com/plone/plone.restapi/pull/1304#pullrequestreview-906965320)?

tisto commented:
@rpatterson @alecghica @avoinea IMHO this is something we should rather discuss face-to-face at the next @plone/volto-team meeting.
